### PR TITLE
fix: support Set<(T, U)> in for-loop multi-variable destructure (#1350)

### LINKED ISSUE #1350: for-loop destructuring fails on Set<(T, U)> — routed through list-only path
## 概要

`for a, b in setOfTuples:` が `error: for loop destructuring requires a list of tuples` で失敗する。`codegen_stmt_loop.cpp::emitForStmt` の多変数バインディング path が map 以外のイテラブルをすべて list として扱い、tuple 要素の Set を拒否している。

## 再現

\`\`\`ry
s: Set<(int, int)> = {(1, 2), (3, 4)}
for a, b in s:
    print(a)
    print(b)
\`\`\`

**実際の動作**:
\`\`\`
error: for loop destructuring requires a list of tuples
  --> repro.ry:2:13
  |
2 | for a, b in s:
  |             ^ for loop destructuring requires a list of tuples
\`\`\`

**期待動作**: 各要素 \`(1, 2)\` \`(3, 4)\` を \`a\`, \`b\` に分解して順に print する。

## 根本原因

\`src/codegen_stmt_loop.cpp:170-177\`:

\`\`\`cpp
if (forBindingArity(s->binding) > 1) {
    llvm::Type *keyTy = getMapKeyType(iterable);
    llvm::Type *valTy = getMapValueType(iterable);
    if (!keyTy || !valTy) {
        llvm::Type *elemTy = getListElementType(iterable);  // Set では null
        auto *structTy = llvm::dyn_cast_or_null<llvm::StructType>(elemTy);
        if (!structTy)
            codegenError("for loop destructuring requires a list of tuples");
        ...
    }
}
\`\`\`

Set iteration は \`getListElementType\` を返さないため、tuple-valued Set がこの早期エラーで止まる。

## 修正方針

\`forBindingArity > 1\` の分岐で:
1. Map path を試す (既存)
2. List path (\`isListType\` 相当で guard) を試す
3. 上記に該当しなければ、Set の tuple 要素 destructuring path にフォールスルー

## 追加コンテキスト

- CodeRabbit (PR #1346 post-hotfix review, 2026-04-24) による指摘
- commit 5284af91 (2026-03-28) 由来の既存バグ
- hotfix PR #1348 スコープ外
- v0.0.13 release blocker ではない (tuple-valued set destructuring は edge case)

### PR DESCRIPTION
**Pull request by t0k0sh1**
## Summary

- `for a, b in setOfTuples:` previously failed with `"for loop destructuring requires a list of tuples"`. The multi-variable binding path in `src/codegen_stmt_loop.cpp` treated every non-map iterable as a list, so `Set<(T, U)>` was rejected before element lookup.
- Mirror the single-variable Set-first / List-fallback pattern (`getSetElementType` + `setHeaderTy_`, falling back to `getListElementType` + `listHeaderTy_`). `List` and `Set` headers share the `{len, cap, data*}` prefix, so the existing GEP arithmetic is reused once `headerTy` is picked. Snapshot `TypeMeta` is routed to `SetElem` vs `ListElem`, and `set_elem_type_name` is preferred when reading the source-level tuple name for `emitForBindingPattern`.

### Scope expansion: `codegen_stmt.cpp` Set annotation path

The `Set<X>` annotation path in `emitVarDecl` previously narrowed the stored inner type to `{List|Map|Set|FnType}`, dropping tuple/enum/record/alias inner types. This PR mirrors the `List<X>` denylist (#1320) so `Set<(str, List<int>)>` annotations actually carry `"(str, List<int>)"` down to the for-loop — otherwise `sum(v)` inside the destructured body fails even after the codegen_stmt_loop fix. **This is wider than the original bug**: named non-primitive inner types (records, enums, aliases, tuples) on `Set<T>` annotations are now propagated as well. `str` stays excluded because Set has no `set_elem_is_str` side channel (unlike List<str> via #1266); `TypeMeta::SetElem` already carries the `i8` LLVM type.

## Test plan

- [x] `./build/ry test tests/spec/for_loop_destructure_meta.test.ry` (16/16, including 5 new `#1350` cases: basic / metadata propagation via `sum(v)` / wildcard / 3-tuple / empty Set)
- [x] `./build/ry_tests` (1937 C++ tests)
- [x] `./build/ry test -p` (168 Ry spec files, 0 failures)
- [x] ASan + UBSan clean
- [x] TSan clean (C++ required tests)
- [x] FileCheck 10/10 pass
- [x] libFuzzer parser/json/utf8 @ 60s each — clean

## Docs & artifacts

- `docs/reference/control-flow.md` — Tuple Destructuring section updated ("list or set of tuples") + new `Set<(int, int)>` example.
- `changelog.d/1350-for-destructure-set.md` — `### Fixed` fragment.
- `KNOWLEDGE.md` — new entry documenting both asymmetries (multi-var for-loop path & `Set<X>` annotation inner-type denylist) trailing the List-side equivalents until this PR, plus a follow-up candidate to extract the shared `(elemTy, headerTy)` selection helper.

Closes #1350

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed tuple destructuring in for-loops to support sets: `for a, b in setOfTuples:` no longer errors.
  * Preserved element-type metadata for sets with non-primitive inner types so destructuring and element access work correctly.

* **Documentation**
  * Control-flow docs updated with examples showing tuple destructuring over sets and a note about unspecified iteration order.

* **Tests**
  * Added tests covering multi-variable destructuring and metadata preservation when iterating sets.

### CODE REVIEW

File: src/codegen_stmt.cpp [hunk 1]
```diff
@@ -816,10 +816,18 @@ void CodeGen::emitVarDecl(const std::string &name,
...
+                } else if (inner != "str" && inner != "int" && inner != "float" && inner != "bool") {
+                    // `str` is intentionally excluded: Set has no
+                    // `set_elem_is_str` side channel (unlike List<str> via
+                    // #1266), and TypeMeta::SetElem already carries the i8
+                    // LLVM type. Everything else — collections, records,
+                    // enums, aliases, tuples like "(str, List<int>)" — is
+                    // kept so for-loop destructure can rebuild component
+                    // metadata via propagateTypeMeta / splitTypeArgs.
+                    setn = inner;
+                }
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_stmt.cpp:835]: Good catch — fixed. The empty-Set literal branch at `src/codegen_stmt.cpp` L230-236 still had the old `{List|Map|Set|Fn}` allowlist and was dropping tuple/record/enum/alias inner names asymmetric to the non-empty literal path I updated.

Applied the same denylist pattern (`!= "str"/"int"/"float"/"bool"`) with `isFunctionTypeName` taking precedence, matching the non-empty path. Rationale for excluding `str` is identical to the non-empty path (Set has no `set_elem_is_str` side channel).

Also added a regression test `"should preserve Set<(str, List<int>)> annotation metadata across empty-literal + add + destructure (#1350 CodeRabbit)"` in `tests/spec/for_loop_destructure_meta.test.ry` that exercises this exact scenario: declare `Set<(str, List<int>)> = {}`, `add()` elements, then destructure with `sum(v)` — this would fail with "sum() requires a list" without the L230-236 fix because the tuple type name never reaches `emitForBindingPattern`. Test passes now; will be included in the next push.
